### PR TITLE
Fix number of patch check for different vision feature select strategy

### LINKED ISSUE #32395: Llava-Next with vision_feature_select_strategy == "full" returns error image size mismatch
### System Info

transformers==4.42.3

### Who can help?

@zucchini-nlp 

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Working example: using `default` vision_feature_select_strategy

```python
from transformers.models.llava_next import LlavaNextForConditionalGeneration, LlavaNextProcessor
from PIL import Image
import requests
import torch

model = LlavaNextForConditionalGeneration.from_pretrained("llava-hf/llava-v1.6-vicuna-7b-hf").to(dtype=torch.bfloat16, device=torch.device("cuda"))
processor = LlavaNextProcessor.from_pretrained("llava-hf/llava-v1.6-vicuna-7b-hf")

url = "http://images.cocodataset.org/val2017/000000039769.jpg"
image_cats = Image.open(requests.get(url, stream=True).raw)

prompt = "USER: <image>\nWhat is shown in this image? ASSISTANT:"

input = processor(prompt, image_cats, padding=True, return_tensors="pt").to("cuda")
output = model.generate(**input, max_length=128)
processor.decode(output[0], skip_special_tokens=True)
```
```
'USER: \nWhat is shown in this image? ASSISTANT: The image shows two cats lying on a pink surface, which appears to be a couch or a bed. The cats are resting with their eyes closed, suggesting they are either sleeping or very relaxed. There are also two remote controls placed near the cats, which might indicate that the cats are in a living room or a space where people watch television. The cats have different patterns on their fur, with one being a tabby and the other a calico.'
```

### Buggy example with `full` vision_feature_select_strategy
```python
from transformers.models.llava_next import LlavaNextForConditionalGeneration, LlavaNextProcessor
from PIL import Image
import requests
import torch

model = LlavaNextForConditionalGeneration.from_pretrained(
    "llava-hf/llava-v1.6-vicuna-7b-hf",
    vision_feature_select_strategy="full"
).to(dtype=torch.bfloat16, device=torch.device("cuda"))
processor = LlavaNextProcessor.from_pretrained("llava-hf/llava-v1.6-vicuna-7b-hf")

url = "http://images.cocodataset.org/val2017/000000039769.jpg"
image_cats = Image.open(requests.get(url, stream=True).raw)

prompt = "USER: <image>\nWhat is shown in this image? ASSISTANT:"

input = processor(prompt, image_cats, padding=True, return_tensors="pt").to("cuda")
output = model.generate(**input, max_length=128)
processor.decode(output[0], skip_special_tokens=True)
```
```
File /opt/conda/lib/python3.10/site-packages/transformers/models/llava_next/modeling_llava_next.py:662, in LlavaNextForConditionalGeneration.pack_image_features(self, image_features, image_sizes, image_newline)
height = width = self.config.vision_config.image_size // self.config.vision_config.patch_size
if height * width != base_image_feature.shape[0]:
    raise ValueError("The number of patches is not consistent with the image size.")

ValueError: The number of patches is not consistent with the image size.
```

> Side note: there is another bug. If you pass `vision_feature_select_strategy` to `.generate()` function, it is not passed to `LlavaNextForConditionalGeneration.forward()` and it uses `default` strategy.
> ```python
> model = LlavaNextForConditionalGeneration.from_pretrained("llava-hf/llava-v1.6-vicuna-7b-hf").to(dtype=torch.bfloat16, > device=torch.device("cuda"))
> input = processor(prompt, image_cats, padding=True, return_tensors="pt").to("cuda")
> output = model.generate(**input, vision_feature_select_strategy="full", max_length=128)
> ```
> The example above "seems" to use `full` strategy, but it is actually using `default` strategy and returns no error, which is not intended.

Back to the main bug example, the reason of such error is due to the following lines:
https://github.com/huggingface/transformers/blob/c1aa0edb48217f416f4bbe6e3a9db1500284513b/src/transformers/models/llava_next/modeling_llava_next.py#L789-L792

Here, `selected_image_feature` shape before applying the select strategy is `[5, 577, 1024]`. If we use `default` select strategy one feature is removed from the second dimension, resulting in `[5, 576, 1024]` feature dimension, which can pass the condition in `self.pack_image_features()`:
https://github.com/huggingface/transformers/blob/c1aa0edb48217f416f4bbe6e3a9db1500284513b/src/transformers/models/llava_next/modeling_llava_next.py#L662-L663
with `height=width=24` thus `height * weight = 576`.

However, if `full` select strategy is used we have the feature with unchanged dimension `[5, 577, 1024]` which raises `ValueError` in `self.pack_image_features()` as `height * width == 576 != 577`.

### Expected behavior

Runs without an error, potentially having a different output from that with `default` strategy.

### PR DESCRIPTION
**Pull request by insujang**
# What does this PR do?

Fixes #32395 (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

I think a new test for this change needs to be added (using `parameterized` to test different `vision_feature_select_strategy` and `vision_feature_layer`), however, I have no idea how the new test would look like. With this change there will be no errors during forward, but do we also need to check the outputs with expected ones? Which test class would be suitable to have this test? @zucchini-nlp  Appreciate it if you could share your thoughts regarding this. Thanks!

## Who can review?

@zucchini-nlp 

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by zucchini-nlp — COMMENTED** [review]
Cool, thanks for working on this!

Yes, we'd need tests and you can add them in https://github.com/huggingface/transformers/blob/e0d82534cc95b582ab072c1bbc060852ba7f9d51/tests/models/llava_next/test_modeling_llava_next.py#L211

Basically we'll run a tiny model with dummy inputs, and verify that changing `select_strategy` run without errors. I don't think we need slow tests here, just make sure after adding the test to run

 `RUN_SLOW=1 pytest tests/models/llava_next/test_modeling_llava_next.py`

and check that all tests pass. If something doesn't pass, just let me know. Might be numerical precision error from different hardware as the PR doesn't change the model logic

**Comment by zucchini-nlp** [comment]
@insujang hey, is the PR in progress or do you need help with adding tests? I can do the test part if you are busy :)

**Comment by insujang** [comment]
Hi @zucchini-nlp , I am sorry but I am too busy to work on it :( I would sincerely appreciate it if you could do. Thank you!

**Comment by zucchini-nlp** [comment]
@insujang no worries, thanks a lot for contribution. Will add a test soon and merge it to `main` 🤗

**Review by zucchini-nlp — APPROVED** [review]
Perfect, LGTM! ❤️ 

I added a test as @insujang is busy and now this can be reviewed

File: src/transformers/models/llava_next/modeling_llava_next.py [hunk 1]
```diff
@@ -648,7 +648,7 @@ def _merge_input_ids_with_image_features(
 
         return final_embedding, final_attention_mask, position_ids, final_labels, final_input_ids
 
-    def pack_image_features(self, image_features, image_sizes, image_newline=None):
+    def pack_image_features(self, image_features, image_sizes, vision_feature_select_strategy, image_newline=None):
```
Comment (LysandreJik) [inline src/transformers/models/llava_next/modeling_llava_next.py:651]: Should it have `"default"` as a default kwarg value so as to not change the signature and keep usages of `pack_image_features` outside of the model working?

(Not necessary if this method should be considered internal)

**Review by LysandreJik — APPROVED** [review]
Feel free to merge if you disapprove the comment below

File: src/transformers/models/llava_next/modeling_llava_next.py (hunk 1 above)
Comment (zucchini-nlp) [reply src/transformers/models/llava_next/modeling_llava_next.py:651]: yes, should be internal as it doesn't make much sense outside of LLaVA

**Pull request merged**